ffi/diplomat/src/errors.rs

```diff
@@ -0,0 +1,147 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use self::ffi::ICU4XError;
+use core::fmt;
+use fixed_decimal::Error as DecimalError;
+use icu_decimal::FixedDecimalFormatError;
+use icu_plurals::PluralRulesError;
+use icu_properties::PropertiesError;
+use icu_provider::{DataError, DataErrorKind};
+
+#[diplomat::bridge]
+pub mod ffi {
+    use alloc::boxed::Box;
+
+    #[repr(C)]
+    /// A common enum for errors that ICU4X may return, organized by API
+    ///
+    /// The error names are stable and can be checked against as strings in the JS API
+    pub enum ICU4XError {
+        // general errors
+        /// The error is not currently categorized as ICU4XError.
+        /// Please file a bug
+        UnknownError = 0,
+        /// An error arising from writing to a string
+        /// Typically found when not enough space is allocated
+        /// Most APIs that return a string may return this error
+        WriteableError = 1,
+
+        // Some input was out of bounds
+        OutOfBoundsError = 2,
+
+        // general data errors
+        // See DataError
+        DataMissingResourceKeyError = 10,
+        DataMissingVariantError = 11,
+        DataMissingLocaleError = 12,
+        DataMissingResourceOptionsError = 13,
+        DataNeedsVariantError = 14,
+        DataNeedsLocaleError = 15,
+        DataExtraneousResourceOptionsError = 16,
+        DataFilteredResourceError = 17,
+        DataMismatchedTypeError = 18,
+        DataMissingPayloadError = 19,
+        DataInvalidStateError = 20,
+        DataCustomError = 21,
+        DataIoError = 22,
+        DataUnavailableBufferFormatError = 23,
+
+        // locale and data struct errors
+        /// The subtag being requested was not set
+        LocaleUndefinedSubtagError = 31,
+        /// The locale or subtag string failed to parse
+        LocaleParserError = 32,
+        /// Attempted to construct an invalid data struct
+        DataStructValidityError = 33,
+
+        // property and decimal errors
+        PropertyUnknownScriptIdError = 40,
```

Comment:
Thought: We should have a discussion about where to draw the line when introducing new entries in the big error enum versus representing them as existing entries.

Reply:
Yeah we definitely do.